posts/2020-10-02-the-tale-about-hibernate-and-temporal-tables.adoc

```diff
@@ -0,0 +1,149 @@
+----
+The result will return the following (which completely makes sense):
+[source,java]
+----
+1 test1 
+1 updated
+----
+Then we can create an Entity to easily operate with the data:
+[source,java]
+----
+@javax.persistence.Entity
+@Table(name = "DepartmentHistory")
+public class DepartmentHistory {
+
+    @Id
+    @Column(name = "DeptId")
+    private int DeptId;
+
+    @Column(name="DeptName")
+    private String DeptName;
+
+// other fields and getters and setters…
+
+}
+----
+and map our result properly to our entity:
+[source,java]
+----
+Query q = entityManager
+    .createNativeQuery("select * from DepartmentHistory where DeptId = 1", 
+    DepartmentHistory.class);
+----
+We expect to get the same result, but this is what we get instead:
+[source,java]
+----
+1 test1
+1 test1
+----
+So we get two rows, but they contain the same values. If there were more rows, they would all return the same data. This happens because of hibernate First level caching. Because all rows have the same ID, when we execute this query, Hibernate caches result on the first row and reuses it for the rest of the rows. 
+
+First level caching can not be turned off, there is nothing "nice" to solve this issue and have Hibernate properly working with temporal tables.
+
+Although…
+
+*…here is how we can trick Hibernate:*
+[source,java]
+----
+Query q = entityManager.createNativeQuery("SELECT ROW_NUMBER() OVER (ORDER BY deptid ASC) AS DeptId, * from DepartmentHistory where DeptId = 1",
+        DepartmentHistory.class);
+----
+We can provide Hibernate with another (fake) DeptId which is in fact the row count. This value will be different for each row and hibernate won't be reusing the result that it cached for the first row, instead it will fetch every row. Importantly, the “fake” DeptId should go first, so the real DeptId will be ignored during conversion of result to our entity. So, when we run it like that, the result is:
+[source,java]
+----
+1 test1
+2 updated
+----
+If we run this query without hibernate, this is what we will see:
+
+image::../media/2020-10-02-the-tale-about-hibernate-and-temporal-tables/db_result_3.png[]
+
+There is a duplicate column for DeptId, and only the first one is used by Hibernate.
+
+== Conclusion
+
+Of course, the result is still not the cleanest, we still have fake IDs instead of the original ones, but this is how you can trick Hibernate without changing your domain. This solution makes sense only if we don’t care about the original ID (in our example we filter by original ID so we already know it anyway). If we select everything from the history table without filtering by ID, the result will be a complete mess:
+[source,java]
+----
+1 test1
+2 updated
+3 test2
+----
+There is no easy trick to go around this without changing your domain. But if you change your entity class for this specific case, it may be a problem for normal queries, just keep that in mind.
```

Comment:
```suggestion
There is no other easy 'trick' to get around this without changing your domain. But if you do the latter for this specific case, it may be a problem for normal queries, just keep that in mind.
```